### PR TITLE
feat: flood the discord channel with latest tweets #319

### PR DESCRIPTION
**Pull request by tgmarinho**
Solve this https://github.com/sibelius/ccsseraphini/issues/22

Cronjob grabs every 1 minute all the latest tweets from `cc/api/tweets`  and use webhook from discord to send the tweets in the #tweets sibs discord channel.

__

Please @sibelius create an account for you here: https://console.cron-job.org/

Then create a job like img bellow, replacing the URL by:

```
https://sseraphini.cc/api/discord/flood
```

Screen Shot 2022-03-25 at 16 36 45

Add to vercel the `.env` WEBHOOK_DISCORD with that link you sent me in DM.

So, Everything should ok!

What is missing:

- [ ] tests
- [ ] auth method on cron job, I can help on it
- [ ] improve how not to repeat the tweets
- [ ] remove comments TODO inside of the code

### CODE REVIEW

File: apps/web/.env.example [hunk 1]
```diff
@@ -3,4 +3,5 @@ TWITTER_BEARER_TOKEN=
 TWITTER_CLIENT_ID=
 TWITTER_CLIENT_SECRET=
 TWITTER_PROFILE_ID=#2313873457
-TWITTER_PROFILE_USER=#sseraphini
\ No newline at end of file
+TWITTER_PROFILE_USER=#sseraphini
+WEB_HOOK_DISCORD= #get this url creating a channel then create a webhook, copy the url and paste here
```
Comment (sibelius) [inline apps/web/.env.example:7]: WEBHOOK_DISCORD ?

File: apps/web/src/pages/api/discord/flood.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+// Learn Discord Webhook: https://discord.com/developers/docs/resources/webhook#execute-webhook
+
+const floodDiscordChannelWithTweets = async (
+  req: NextApiRequest,
+  res: NextApiResponse,
+) => {
+  // TODO add auth method, only cron job can access this endpoint
+  // TODO handle with discord rate limit
+  // cron will trigger the API every 1 minute
+  // get the lastest tweets
```
Comment (sibelius) [inline apps/web/src/pages/api/discord/flood.ts:13]: what about adding a .env for this?

File: apps/web/src/pages/api/discord/flood.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  const url = `${httpProtocol}://${req.headers.host}/api/tweets?query=${query}`;
+
+  const response = await fetch(url);
+
+  if (response.status !== 200)
+    res.status(500).send("I can't get the tweets, try again later");
+
+  const data = await response.json();
+
+  const WEB_HOOK = process.env.WEB_HOOK_DISCORD as string;
```
Comment (sibelius) [inline apps/web/src/pages/api/discord/flood.ts:31]: move WEBHOOK_DISCORD to config.ts

File: apps/web/src/pages/api/discord/flood.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      method: 'POST',
+      body: JSON.stringify({
+        content: `https://twitter.com/${tweet.userInfo.username}/status/${tweet.id}`,
+      }),
+    });
+  });
+
+  try {
+    // TODO handle with discord rate limit
+    await Promise.all(promises);
```
Comment (sibelius) [inline apps/web/src/pages/api/discord/flood.ts:49]: avoid promise all

prefer

for of

**Review by sibelius — APPROVED** [review]
cool

File: apps/web/src/pages/api/discord/flood.ts (hunk 2 above)
Comment (tgmarinho) [reply apps/web/src/pages/api/discord/flood.ts:13]: this what? @sibelius

File: apps/web/src/pages/api/discord/flood.ts (hunk 3 above)
Comment (tgmarinho) [reply apps/web/src/pages/api/discord/flood.ts:31]: ok

File: apps/web/src/pages/api/discord/flood.ts (hunk 4 above)
Comment (tgmarinho) [reply apps/web/src/pages/api/discord/flood.ts:49]: ok

File: apps/web/.env.example (hunk 1 above)
Comment (tgmarinho) [reply apps/web/.env.example]: ok

**Comment by tgmarinho** [comment]
To solve security you can set headers: "header-flood-secret-key": "my-super-passwd" and store inside of the .env `CRONJOB_HEADER_SECRET =  "my-super-passwd"` 

and compare with the value of "header-flood-secret-key" === CRONJOB_HEADER_SECRET

**Comment by tgmarinho** [comment]
To fix duplicated, cron job can call every 5 minutes, but we can loss some tweets or we can store the latest tweet id sent do discord, maybe using mongoatlas or faunadb

File: apps/web/src/pages/api/discord/flood.ts (hunk 2 above)
Comment (sibelius) [reply apps/web/src/pages/api/discord/flood.ts]: To validate the auth part

**Pull request merged**